zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Leader.java

```diff
@@ -240,16 +249,47 @@ public void addLearnerHandler(LearnerHandler learner) {
      * @param peer
      */
     @Override
-    public void removeLearnerHandler(LearnerHandler peer) {
-        synchronized (forwardingFollowers) {
-            forwardingFollowers.remove(peer);
-        }
-        synchronized (learners) {
-            learners.remove(peer);
-        }
-        synchronized (observingLearners) {
-            observingLearners.remove(peer);
-        }
+    public synchronized void removeLearnerHandler(LearnerHandler peer) {
+        forwardingFollowers.remove(peer);
+
+        /*
+        * ZOOKEEPER-3922
+        *
+        * We will need to re-validate the outstandingProposal to maintain the progress of ZooKeeper.
+        * It is likely a proposal is waiting for enough ACKs to be committed. The proposals are sent out, but the
+        * only follower goes away which makes the proposals will not be committed until the follower recovers back.
+        * An earlier proposal which is not committed will block any further proposals. So, We need to re-validate those
+        * outstanding proposal with the help from Oracle. A key point in the process of re-validation is that the proposals
+        * need to be processed in order.
+        *
+        * We make the whole method blocking to avoid any possible race condition on outstandingProposal and lastCommitted
+        * as well as to avoid nested synchronization.
+        *
+        * As a more generic approach, we pass the object of forwardingFollowers to QuorumOracleMaj to determine if we need
+        * the help from Oracle.
+        *
+        *
+        * the size of outstandingProposals can be 1. The only one outstanding proposal is the one waiting for the ACK from
+        * the leader.
+        * */
+        self.getQuorumVerifier().updateNeedOracle(new ArrayList<>(forwardingFollowers));
```

Comment:
this method has a lot of synchronization and ordering. is this exactly where this code needs to be? it would be nice to move this to a method that gets called only if we are using a oracle verifier.